src/lsm/binary_search.zig

```diff
@@ -18,7 +18,7 @@ pub fn binary_search_values_raw(
     values: []const Value,
     key: Key,
 ) u32 {
-    assert(values.len > 0);
+    if (values.len == 0) return 0;
```

Comment:
The assertion here is actually deliberate (albeit surprising). It forces the caller to specifically handle the empty case.

Reply:
Though I don't remember why :wink:. @jorangreef would know.

Reply:
My instinct was that the caller always has to handle the case where result == values.len anyway, but maybe there are bad cases I haven't thought of yet.

Reply:
Yes, the assertion is to document that within the `lsm` namespace we expect to only binary search values in a table, where tables always contain values. This means we don't require an optional return value. i.e. We can always return an index to a key.

Reply:
The `result == values.len` case also shouldn't happen, because we first check the `key_min` and `key_max` of a table to see if the table is in range. We shouldn't be binary searching within a table if the result is going to be outside and to the right of the range.

Reply:
We use binary search in other contexts in lsm besides searching tables (`SegmentedArray`, `ManifestLevel`).

Reply:
Ah, good point. We should still always expect to be searching for `values.len > 0` there though?

Reply:
When we are searching for an insertion index, we need to manually check for an empty array first, e.g.: https://github.com/tigerbeetledb/tigerbeetle/blob/29f4dfd758068e4c7d0b72d61917146bc735738f/src/lsm/manifest_level.zig#L580-L583

Reply:
Would you prefer an optional result signature then for binary search to return null `if values.len == 0`?

Then we can assert on this by `.?` when searching tables.

Reply:
:thinking: Would it be more consistent to return `0`, to match the case where the key is greater than all keys in the list and it returns `values.len`?

Reply:
It should actually always return a valid index to the next greatest key. At least, that's the documentation there. That's how we used it also I think in the old ManifestLevel.

Reply:
Returning 0 and exact=false does match the invariants. It's also correct for the case where we're trying to insert something into a sequence, regardless of whether the sequence is currently empty.

(If we don't put the assert back in, I should go and replace the usages that special-case empty slices)

Reply:
We wrote concurrent and disagreeing comments :D

Reply:
...of which yours is correct, on further reading.

Reply:
We currently have:

`.exact = index < values.len and compare_keys(key_from_value(&values[index]), key) == .eq,`

Which is misleading if we guarantee `index < values.len` already.

Reply:
If we return the last index, rather than the end-of-the-array index in the case where the needle is greater than everything in the slice, then I think https://github.com/tigerbeetledb/tigerbeetle/blob/29f4dfd758068e4c7d0b72d61917146bc735738f/src/lsm/segmented_array.zig#L612 is unreachable?

Reply:
I don't think there are any other cases that do anything with the index if exact == false.

Reply:
Agreed with you on all three points. :D

We should clean this up—at present it reflects the evolution of how binary search was introduced, and it's not consistent at all.

Your interface would be perfect.

Reply:
On stepping through, I think @jorangreef is right that the documentation and code match for search_raw. I should have read that closely before changing the behavior

I find the documented behavior surprising though and the code in segmented_array at least seems to assume the behavior that I expect for binary_search - to return the insertion point if there is no exact match.

Reply:
...github needs a typing indicator :D

Reply:
@jorangreef is typing...

Reply:
> I find the documented behavior surprising though and the code in segmented_array at least seems to assume the behavior that I expect for binary_search - to return the insertion point if there is no exact match.

Yeah, I didn't read the binary search docs closely enough when updating the segmented array :sweat_smile:.

Reply:
Ok, todo for me next week:
* [ ] Change the documentation and tests to match the usual behavior of binary search.
* [ ] Review all usages to make sure no correct behavior changes.
* [ ] Test if the current usage in segmented arrays is correct.

Reply:
Let's fix it to be non-surprising!

Reply:
Just for a bonus gotcha, segmented array also has an inlined implementation of binary search (in `search()`) to iterate over nodes, before it drops into `binary_search_values_raw()` to search within the found node. :grin:

Reply:
I added a lot of config.verify -only asserts and tweaked the explict tests to cover edge cases. It seems that @sentientwaffle's expectations were accurate and the documentation on binary_search_raw is wrong. So I didn't have to change the implementation itself.